### ISSUE TITLE
Alias type conditions to desctrutured variables

### ISSUE THREAD

**Issue by ilogico** [microsoft/TypeScript#46143]
# Suggestion

## 🔍 Search Terms
is:issue is:open control flow analysis destructure 

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

TypeScript 4.4 has a feature called [Control Flow Analysis of Aliased Conditions and Discriminants](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#cfa-aliased-conditions), which essentially allows a variable to carry type information about another variable.
I believe something similar could happen when destructuring discriminated unions.

Object destructuring is a very common pattern in JavaScript, but we need to avoid it often in TypeScript because it loses type information:

```ts
declare var it: Iterator<number>;
const { value, done } = it.next();
if (!done) {
  value; // we know it's a number, but it's any
}
```

has to be artificially transformed to satisfy the typechecker:

```ts
const result = it.next();
if (!result.done) {
  result.value; // is narrowed to a number
}
```

With TypeScript 4.4, it looks like we're very close to having this feature:
```ts
const result = it.next();
const { value, done } = result;
if (!done) {
    result.value; // `done` tracks result's type
    value; // but not value's
}
```
## 📃 Motivating Example
These patterns emerge naturally in JavaScript and currently have to be transformed to fit TypeScript's ability to infer types.
The IteratorResult example above.
Another example is with action to state reducers:
```ts
type Action =
  | { type: 'A', payload: number }
  | { type: 'B': payload: string };

function reducer(state: Whatever, { type, payload }: Action) {
  switch (type) {
    case 'A':
      payload.toFixed();
      break;
    case 'B':
      payload.toUpperCase();
      break;
  }
}
```
## 💻 Use Cases

I think the examples above already cover this.

**Comment by fatcerberus** [comment]
This looks similar in spirit to https://github.com/microsoft/TypeScript/issues/35873

**Comment by ilogico** [comment]
@fatcerberus, but maybe not as ambitious.
Anyway, I'm aware this should've been requested/reported multiple times.
But because TS4.4 brought a feature somewhat related to this, I thought I could use the opportunity to bring this up.

**Comment by IllusionMH** [comment]
This case was explicitly mentioned in PR that implemented aliases #44730 (at the end, before list if fixed issues)
However not sure if there is big difference with #35873 or if there is better issue tracking correlation during destructuring only.
